### PR TITLE
Consume device aggregator output from socket over http, tra…

### PR DESCRIPTION
**Pull request by tanabarr**
…nsform into agent output format and feed into host/manager side linux plug in handler.

The device aggregator will hold details of devices connected to each host running an IML agent, in the reactive architecture we want to consume from the aggregator rather than using the agent plug-in polling mechanism.

Remove logic for gathering device information in Linux plug-in over cli, instead normalise output of device aggregator in the manager side of the plug-in to transform it into expected agent output format (for now).
Still using the following mechanism of the plug-in manager to check aggregate output for changes at intervals.

Fix passing nested dataset structures into resource manager by reflecting new format from `device-scanner`.

This PR additionally integrates #337 #131 #372 as pre-cursive work to move device processing code _from_ the agent linux device plug-in component _to_ the manager component.

previous in stack <master> (this is the bottom)
next in stack #497 

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

---
This change is [Reviewable](https://reviewable.io/reviews/intel-hpdd/intel-manager-for-lustre/456)

### CODE REVIEW

File: chroma-manager/chroma_core/plugins/block_devices.py [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+    zpools = {}
+    datasets = {}
+    # fixme: get zvols from zfs_objs using some identifier to discern from datasets
+
+    for pool in zpool_map.values():
+        name = pool['name']
+        guid = pool['guid']
+
+        size = pool['size']
+        drive_mms = get_drives([child['Disk'] for child in pool['vdev']['Root']['children']],
```
Comment (jgrund) [inline chroma-manager/chroma_core/plugins/block_devices.py]: This won't be sufficient for mapping all the drives. 

The `VDev` tree is recursive and can contain things like mirrors.

File: chroma-agent/chroma_agent/action_plugins/detect_scan.py [hunk 2]
```diff
@@ -15,8 +15,10 @@
...
     '''
-    Allows local targets to be examined. Not the targets are only examined once with the results cached. Detecting change
-    therefore requires a new instance to be created and queried.
+    Allows local targets to be examined.
+    Note the targets are only examined once with the results cached.
+    Detecting change therefore requires a new instance
+    to be created and queried.
     '''
 
     def __init__(self, target_devices):
```
Comment (jgrund) [inline chroma-agent/chroma_agent/action_plugins/detect_scan.py]: We should have all the info we need in device-scanner / aggregator now.

File: chroma-agent/chroma_agent/device_plugins/linux.py [hunk 3]
```diff
@@ -22,84 +14,9 @@ class LinuxDevicePlugin(DevicePlugin):
...
-        return self._scan_devices(True)
+        return {}
 
     def update_session(self):
-        trigger_plugin_update = self.trigger_plugin_update
-        self.trigger_plugin_update = False
-        scan_result = self._scan_devices(trigger_plugin_update)
-
-        return scan_result
+        return {}
```
Comment (jgrund) [inline chroma-agent/chroma_agent/device_plugins/linux.py]: given this is now basically a no-op, any reason we can't drop the whole plugin?

File: chroma-agent/chroma_agent/device_plugins/linux_components/block_devices.py [hunk 4]
```diff
@@ -90,11 +95,20 @@ def as_device(x):
...
         'device_type': x.get('DEVTYPE'),
         'device_path': x.get('DEVPATH'),
         'partition_number': x.get('ID_PART_ENTRY_NUMBER'),
         'is_ro': x.get('IML_IS_RO'),
-        'parent': None
+        'parent': None,
+        'dm_multipath': x.get('DM_MULTIPATH_DEVICE_PATH'),
+        'dm_lv': x.get('DM_LV_NAME'),
+        'dm_vg': x.get('DM_VG_NAME'),
+        'dm_uuid': x.get('DM_UUID'),
```
Comment (jgrund) [inline chroma-agent/chroma_agent/device_plugins/linux_components/block_devices.py]: Is this code still relevant?

I removed dm_uuid in favor of vg_uuid and lv_uuid I think.

File: chroma-agent/chroma_agent/device_plugins/linux_components/block_devices.py [hunk 5]
```diff
@@ -126,76 +140,314 @@ def filter_device(x):
...
+    return DeviceMaps(info["blockDevices"], info["zpools"])
+
+
+def create_device_list(device_dict):
+    return pipe(device_dict.itervalues(),
                 cmap(as_device), cfilter(filter_device), list)
 
 
-def add_to_ndp(xs, ys):
+def parse_lvm_uuids(dm_uuid):
```
Comment (jgrund) [inline chroma-agent/chroma_agent/device_plugins/linux_components/block_devices.py]: This code is dead now.

File: chroma-agent/chroma_agent/device_plugins/linux_components/block_devices.py [hunk 6]
```diff
@@ -236,8 +491,30 @@ def build_paths(x):
...
+    c_path_to_major_minor = path_to_major_minor(node_block_devices, ndt)
+
+    return pipe(device_paths,
+                cmap(c_path_to_major_minor), cfilter(None), list)
+
+
+@curry
+def path_to_major_minor(node_block_devices, ndt, device_path):
+    """ Return device major minor for a given device path """
+    return node_block_devices.get(ndt.normalized_device_path(device_path))
```
Comment (jgrund) [inline chroma-agent/chroma_agent/device_plugins/linux_components/block_devices.py]: Reading through this file, I assume this goes away in later commits.

File: chroma-agent/chroma_agent/device_plugins/linux_components/device_mapper.py [hunk 7]
```diff
@@ -1,266 +0,0 @@
-# Copyright (c) 2017 Intel Corporation. All rights reserved.
```
Comment (jgrund) [inline chroma-agent/chroma_agent/device_plugins/linux_components/device_mapper.py]: yay

File: chroma-manager/chroma_core/plugins/block_devices.py [hunk 8]
```diff
@@ -0,0 +1,559 @@
+# Copyright (c) 2017 Intel Corporation. All rights reserved.
+# Use of this source code is governed by a MIT-style
+# license that can be found in the LICENSE file.
+from collections import defaultdict
```
Comment (jgrund) [inline chroma-manager/chroma_core/plugins/block_devices.py]: Assuming this is mostly gone in a later commit as well.

File: chroma-manager/chroma_core/plugins/linux.py [hunk 9]
```diff
@@ -175,23 +181,46 @@ def __init__(self, resource_manager, scannable_id=None):
...
         initiate_device_poll = False
         reported_device_node_paths = []
 
-        for expected_item in ['vgs', 'lvs', 'emcpower', 'zfspools', 'zfsdatasets', 'zfsvols', 'mpath', 'devs', 'local_fs', 'mds']:
-            if expected_item not in devices:
+        # todo: error handling
+        fqdn = ManagedHost.objects.get(id=host_id).fqdn
+        devices = get_block_devices(fqdn)
+
+        # todo: EMCPower Device detection has been deprecated and mpath it is not provided and is unused
```
Comment (jgrund) [inline chroma-manager/chroma_core/plugins/linux.py]: Mpath still needed, assume it's added back in later commits.

File: chroma-manager/chroma_core/plugins/linux.py [hunk 10]
```diff
@@ -366,32 +396,30 @@ def preferred_serial(bdev):
...
-        self._map_drives_to_device_to_node(devices, host_id, 'emcpower', EMCPower, [], reported_device_node_paths)
+        # self._map_drives_to_device_to_node(devices, host_id, 'emcpower', EMCPower, [], reported_device_node_paths)
 
+        log.debug("zfspools: {}".format(devices['zfspools']))
         initiate_device_poll = self._map_drives_to_device_to_node(devices, host_id, 'zfspools', ZfsPool, ['name'], reported_device_node_paths) or initiate_device_poll
 
         initiate_device_poll = self._map_drives_to_device_to_node(devices, host_id, 'zfsdatasets', ZfsDataset, ['name'], reported_device_node_paths) or initiate_device_poll
 
-        initiate_device_poll = self._map_drives_to_device_to_node(devices, host_id, 'zfsvols', ZfsVol, ['name'], reported_device_node_paths) or initiate_device_poll
+        # initiate_device_poll = self._map_drives_to_device_to_node(devices, host_id, 'zfsvols', ZfsVol, ['name'], reported_device_node_paths) or initiate_device_poll
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply chroma-manager/chroma_core/plugins/linux.py]: addressed in later commits

File: chroma-manager/chroma_core/plugins/linux.py [hunk 11]
```diff
@@ -18,6 +21,9 @@
 from chroma_core.models import VolumeNode
 from settings import SERIAL_PREFERENCE
 
+log = log_register('plugin_runner')
+log.setLevel(DEBUG)
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply chroma-manager/chroma_core/plugins/linux.py]: yes just for the moment

File: chroma-manager/chroma_core/plugins/linux.py (hunk 9 above)
Comment (tanabarr) [reply chroma-manager/chroma_core/plugins/linux.py]: yes it is

File: chroma-manager/chroma_core/plugins/block_devices.py (hunk 8 above)
Comment (tanabarr) [reply chroma-manager/chroma_core/plugins/block_devices.py]: yes almost all

File: chroma-agent/chroma_agent/device_plugins/linux_components/block_devices.py (hunk 4 above)
Comment (tanabarr) [reply chroma-agent/chroma_agent/device_plugins/linux_components/block_devices.py]: addressed in later commits

File: chroma-agent/chroma_agent/device_plugins/linux.py (hunk 3 above)
Comment (tanabarr) [reply chroma-agent/chroma_agent/device_plugins/linux.py]: we still currently use the plug-in mechanism to trigger manager side code, that will be removed later

File: chroma-agent/chroma_agent/action_plugins/detect_scan.py (hunk 2 above)
Comment (tanabarr) [reply chroma-agent/chroma_agent/action_plugins/detect_scan.py]: will be replacing/removing both detect_scan and lustre in future commits (still to do)

**Pull request merged**